docs/source/notes/extending.rst

```diff
@@ -14,59 +14,104 @@ Extending :mod:`torch.autograd`
 
 Adding operations to :mod:`~torch.autograd` requires implementing a new
 :class:`Function` subclass for each operation. Recall that :class:`Function` s
-are what :mod:`~torch.autograd` uses to compute the results and gradients, and
-encode the operation history. Every new function requires you to implement 2 methods:
-
-- :meth:`~Function.forward` - the code that performs the operation. It can take
+are what :mod:`~torch.autograd` uses to encode the operation history and compute
+gradients.
+
+When to use
+^^^^^^^^^^^
+In general, implement a custom function if you want to perform computations in your model
+that are not differentiable or rely on non-Pytorch libraries (e.g., NumPy), but
+still wish for your operation to chain with other ops and work with the autograd engine.
+
+In some situations, custom functions can also be used to improve performance and
+memory usage: If you implemented your forward and backward passes using a C++
+extension, you can wrap them in :class:`~Function` to interface with the autograd
+engine. If you'd like to reduce the number of buffers saved for the backward pass,
+custom functions can be used to combine ops together.
+
+When not to use
+^^^^^^^^^^^^^^^
+If you can already write your function in terms of PyTorch's built-in ops, its
+backward graph is (most likely) already able to be recorded by autograd. In this case, you do
+not need to implement the backward function yourself. Consider using a plain
+old Python function.
+
+If you need to maintain state, i.e., trainable parameters, you should (also) use a
+custom module. See the section below for more information on extending :mod:`torch.nn`.
+
+If you'd like to alter the gradients during the backward pass or perform a side
+effect, consider using function or module hooks.
+
+How to use
+^^^^^^^^^^
+Take the following steps:
+1. Subclass :class:`~Function` and implement the :meth:`~Function.forward` and
+:meth:`~Function.backward` methods
+2. Call the proper methods on the `ctx` argument.
+3. Declare whether your function supports double backward
+4. Validate whether your gradients are correct using gradcheck
+
+**Step 1:** After subclassing :class:`Function`, you'll need to define 2 methods:
+
+- :meth:`~Function.forward` is the code that performs the operation. It can take
   as many arguments as you want, with some of them being optional, if you
   specify the default values. All kinds of Python objects are accepted here.
   :class:`Tensor` arguments that track history (i.e., with
   ``requires_grad=True``) will be converted to ones that don't track history
   before the call, and their use will be registered in the graph. Note that this
   logic won't traverse lists/dicts/any other data structures and will only
-  consider :class:`Tensor` s that are direct arguments to the call. You can
+  consider :class:`Tensor`s that are direct arguments to the call. You can
```

Comment:
Note to self: check if the space between the s and class is necessary

Reply:
From the rendered doc https://docs-preview.pytorch.org/60312/notes/extending.html?highlight=custom%20function
looks like it is necessary. However, with the space it looks bad as well, see a couple lines above where we have ":class:`Function` s". So I'm okay with just removing the link in the plural case.

Reply:
Also replacing Tensors -> tensors for consistency (unless we mean to refer to the actual class in which case it would need to be uppercase ":class:`Tensor`".